operator/builtin/parser/csv/csv.go

```diff
@@ -87,42 +94,66 @@ func (c CSVParserConfig) Build(context operator.BuildContext) ([]operator.Operat
 // CSVParser is an operator that parses csv in an entry.
 type CSVParser struct {
 	helper.ParserOperator
-	header         []string
-	fieldDelimiter rune
-	numFields      int
-	lazyQuotes     bool
+	fieldDelimiter  rune
+	header          []string
+	headerAttribute string
+	lazyQuotes      bool
+	parse           parseFunc
 }
 
-// Process will parse an entry for csv.
-func (r *CSVParser) Process(ctx context.Context, entry *entry.Entry) error {
-	return r.ParserOperator.ProcessWith(ctx, entry, r.parse)
-}
-
-// parse will parse a value using the supplied csv header.
-func (r *CSVParser) parse(value interface{}) (interface{}, error) {
-	var csvLine string
-	switch val := value.(type) {
-	case string:
-		csvLine = val
-	default:
-		return nil, fmt.Errorf("type '%T' cannot be parsed as csv", value)
-	}
-
-	reader := csvparser.NewReader(strings.NewReader(csvLine))
-	reader.Comma = r.fieldDelimiter
-	reader.FieldsPerRecord = r.numFields
-	reader.LazyQuotes = r.lazyQuotes
-	parsedValues := make(map[string]interface{})
-
-	record, err := reader.Read()
+type parseFunc func(interface{}) (interface{}, error)
 
-	if err != nil {
-		return nil, err
+// Process will parse an entry for csv.
+func (r *CSVParser) Process(ctx context.Context, e *entry.Entry) error {
+	if r.headerAttribute != "" {
+		h, ok := e.Attributes[r.headerAttribute]
+		if !ok {
+			err := fmt.Errorf("failed to read dynamic header attribute %s", r.headerAttribute)
+			r.Error(err)
+			return err
+		}
+		headers := strings.Split(h, string([]rune{r.fieldDelimiter}))
+		r.parse = generateParseFunc(headers, r.fieldDelimiter, r.lazyQuotes)
 	}
+	return r.ParserOperator.ProcessWith(ctx, e, r.parse)
```

Comment:
Modifying the operator on the fly like this will limit the operator's ability to be used concurrently. Two different entries with different headers would potentially end up using one parse func. It doesn't look like there's any benefit to save the func anyways.

Also, maybe make `generateParseFunc` a receiver on the operator, since we're mostly just passing in a bunch of operator related config values anyways.

```suggestion
	if r.parse != nil {
		return r.ParserOperator.ProcessWith(ctx, e, r.parse)
	}
	
	h, ok := e.Attributes[r.headerAttribute]
	if !ok {
		err := fmt.Errorf("failed to read dynamic header attribute %s", r.headerAttribute)
		r.Error(err)
		return err
	}
	headers := strings.Split(h, string([]rune{r.fieldDelimiter}))
	parse := generateParseFunc(headers, r.fieldDelimiter, r.lazyQuotes)
	return r.ParserOperator.ProcessWith(ctx, e, parse)
```

Reply:
I think that's fair to change it to a receiver on the struct itself. 

Can you elaborate on why this would limit the parsers ability to be used concurrently? Since a parse function is being generated per entry and the it seems like the `ParseOperator` should be able to handle a new parse function every time, it's not obvious to me if we're creating a race condition further down in the stack.

Reply:
> Can you elaborate on why this would limit the parsers ability to be used concurrently?

Ok, roughly the steps to process one entry:
* Split headers
* Generate parse func
* Save parse func onto operator
* Process entry with operator's parse func

So if we have two entries, A & B, processed this way concurrently, they can interleave like:
* [A] Split headers
* [B] Split headers
* [A] Generate parse func
* [A] Save parse func onto operator
* [B] Generate parse func
* [B] Save parse func onto operator  // operator's parse func is now that of entry B
* [A] Process entry with operator's parse func // parsed with wrong parse func
* [B] Process entry with operator's parse func

Reply:
Oh yeah saving the parse function we shouldn't save it in the dynamic case. We should just pass it through. The saving it is an optimization for the non dynamic case. I'll change this. Good call.